feat(performance) - Adding key transactions to event stats

wmak opened the pull request:
- Yanking the key transaction query conditions out into its own function
- Reusing the conditions for event stats
- Adding the same param that forces conditions on a query for timeseries
  too.

wmak committed:
feat(performance) - Adding key transactions to event stats

- Yanking the key transaction query conditions out into its own function
- Reusing the conditions for event stats
- Adding the same param that forces conditions on a query for timeseries
  too.

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -665,6 +727,9 @@ def timeseries_query(selected_columns, query, params, rollup, reference_event=No
         snuba_args["aggregations"][0][2] = "count"
 
+    if conditions is not None:
+        snuba_args["conditions"].extend(conditions)
```
How do we know that aliases have been resolved on these conditions? It might be safer/better to add a `raw_query` call inside `key_transaction_timeseries_query`. That way we can ensure that aliases are always resolved correctly now and in the future.

wmak replied:
🤔 I can't think of a good way to do that without moving the rest of this function into a helper, what about changing the parameter to an optional `key_transaction_queryset` instead so its more explicit?

markstory replied:
I would likely duplicate the guts of this method personally. It is a small amount of duplication to avoid a potentially unsafe interface.

wmak replied:
Didn't end up duplicating, moved most of it to shared function instead. But can switch to duplication if that's not a good idea.

markstory commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -571,3 +574,148 @@ def test_invalid_interval(self):
+                    "start": iso_format(before_now(hours=2)),
+                    "interval": "30m",
+                    "yAxis": "count()",
+                    "keyTransactions": True,
```
Should we have a `stats` endpoint inside the other key-transactions endpoints? We might be overloading this endpoint quite a bit with key transactions, and the topN results that are coming.

wmak replied:
Created a separate endpoint for this, but they're still very similar functionally so I moved a bunch into the api base.

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -611,7 +612,66 @@ def query(
+        user_query,
+        params,
+        rollup,
+        referrer=referrer,
```
It would be good if key transaction queries had a distinct referrer so we can generate aggregates on them in our internal discover reports.

wmak replied:
:+1: Referrer is being set from the API side, but should I move that here instead?

markstory commented on src/sentry/api/bases/organization_events.py:
```diff
@@ -1,18 +1,32 @@
+# Maximum number of results we are willing to fetch.
+# Clients should adapt the interval width based on their
+# display width.
+MAX_POINTS = 4500
```
We have a few copies of this value. Should we consolidate them into either `sentry.constants`, `sentry.snuba.discover` or `sentry.utils.snuba`?

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -727,8 +689,59 @@ def timeseries_query(
+
+def key_transaction_timeseries_query(selected_columns, query, params, rollup, referrer, queryset):
+    snuba_args, snuba_filter = get_timeseries_snuba_vars(selected_columns, query, params, rollup)
+    snuba_args.get("conditions").extend(key_transaction_conditions(queryset))
```
You may hit a conflict here. I recently merged changes to remove `snuba_args` dictionary from most of the discover internals, as `eventstore.Filter` offers similar features without object attributes.

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -727,8 +689,59 @@ def timeseries_query(
+    return snuba_args, snuba_filter
+
+
+def key_transaction_timeseries_query(selected_columns, query, params, rollup, referrer, queryset):
```
Would be good to have a docstring for this.

markstory commented on src/sentry/discover/endpoints/discover_key_transactions.py:
```diff
@@ -98,3 +98,30 @@ def delete(self, request, organization):
+                query=query,
+                params=params,
+                rollup=rollup,
+                referrer="api.organization-event-stats.key-transactions",
```
:+1: Thanks

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -727,8 +689,59 @@ def timeseries_query(
+                    conditions based on the provided reference.
+    referrer (str|None) A referrer string to help locate the origin of this query.
+    conditions (Sequence[any]) List of conditions that are passed directly to snuba without
+                    any additional processing.
```
This line is no longer applicable.

wmak committed:
chore(visibility) - Moving get_rollup to a util function

- Originally wanted to move `MAX_POINTS` to a single spot, and still did
  that, but noticed that get_rollup is shared here between two
  endpoints. Moved it to a util to reduce code repeat.

wmak commented:
WIP: Tests are flakey because of event_stats + transactions
